File: JavaPracticePad/app/src/main/java/com/practice/moviedb/adapters/MovieListAdapter.java
package com.practice.moviedb.adapters;

import android.content.Context;
import android.content.Intent;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.bumptech.glide.Glide;
import com.practice.moviedb.R;
import com.practice.moviedb.activities.MovieActivity;
import com.practice.moviedb.models.Result;
import com.practice.moviedb.models.TopRatedMovie;

import java.util.ArrayList;
import java.util.List;

public class MovieListAdapter extends RecyclerView.Adapter<MovieListAdapter.MovieListViewHolder> {

    private Context context;
    private TopRatedMovie topRatedMovie;
    private List<Result> movieList;

    public MovieListAdapter(Context context) {
        this.context = context;
    }

    @NonNull
    @Override
    public MovieListViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {

        View view = LayoutInflater.from(context).inflate(R.layout.layout_movie_item, parent, false);

        return new MovieListViewHolder(view);
    }

    @Override
    public void onBindViewHolder(@NonNull MovieListViewHolder holder, int position) {
        final Result result = movieList.get(position);

        holder.movieTitleTextView.setText(result.getTitle());
        holder.releasedDateTextView.setText(String.format("Released Date : %s", result.getReleaseDate()));

        Glide.with(context)
                .load("https://image.tmdb.org/t/p/w600_and_h900_bestv2" + result.getPosterPath())
                .placeholder(R.drawable.ic_movie_poster)
                .into(holder.moviePosterImageView);

        holder.itemView.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                Intent intent = new Intent(context, MovieActivity.class);
                context.startActivity(intent);
            }
        });
    }

    @Override
    public int getItemCount() {
        return movieList.size();
    }

    class MovieListViewHolder extends RecyclerView.ViewHolder {

        private ImageView moviePosterImageView;

        private TextView movieTitleTextView;
        private TextView releasedDateTextView;

        MovieListViewHolder(@NonNull View itemView) {
            super(itemView);

            moviePosterImageView = itemView.findViewById(R.id.movie_poster);

            movieTitleTextView = itemView.findViewById(R.id.movie_title);
            releasedDateTextView = itemView.findViewById(R.id.movie_released);
        }
    }

    public void setTopRatedMovie(TopRatedMovie topRatedMovie) {
        this.topRatedMovie = topRatedMovie;

        if (topRatedMovie.getResults() == null) {
            this.movieList = new ArrayList<>();
        } else {
            this.movieList = topRatedMovie.getResults();
        }
    }
}

File: MovieDB-Kotlin/app/src/main/java/com/practice/moviedatabase/adapters/MovieListAdapter.kt
package com.practice.moviedatabase.adapters

import android.content.Context
import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import com.bumptech.glide.Glide
import com.practice.moviedatabase.R
import com.practice.moviedatabase.AllConstants
import com.practice.moviedatabase.adapters.viewholders.MovieEvenListViewHolder
import com.practice.moviedatabase.base.ItemClickListener
import com.practice.moviedatabase.models.Result
import com.practice.moviedatabase.models.TopRatedMovie
import com.practice.moviedatabase.adapters.viewholders.MovieOddListViewHolder
import java.util.*

class MovieListAdapter(private val context: Context) : RecyclerView.Adapter<RecyclerView.ViewHolder>() {

    private val movieOddItem: Int = 0
    private val movieEvenItem: Int = 1

    private var topRatedMovie: TopRatedMovie? = null
    private var movieList: List<Result>? = null

    private lateinit var listener: ItemClickListener

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {

        val oddView = LayoutInflater.from(context).inflate(R.layout.layout_movie_item_odd, parent, false)
        val evenView = LayoutInflater.from(context).inflate(R.layout.layout_movie_item_even, parent, false)

        return when (viewType) {
            movieOddItem -> MovieOddListViewHolder(oddView)
            else -> MovieEvenListViewHolder(evenView)
        }
    }

    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {

        if (holder.itemViewType == movieOddItem) {

            val oddViewHolder: MovieOddListViewHolder = holder as MovieOddListViewHolder
            val result = movieList!![position]

            oddViewHolder.bind(context, result, listener)
        } else {

            val evenViewHolder: MovieEvenListViewHolder = holder as MovieEvenListViewHolder
            val result = movieList!![position]

            evenViewHolder.bind(context, result, listener)
        }
    }

    override fun getItemViewType(position: Int): Int {
        return if (position.and(1) == 1) movieOddItem else movieEvenItem
    }

    override fun getItemCount(): Int {
        return movieList!!.size
    }

    fun setTopRatedMovie(topRatedMovie: TopRatedMovie) {
        this.topRatedMovie = topRatedMovie

        if (topRatedMovie.results == null) {
            this.movieList = ArrayList()
        } else {
            this.movieList = topRatedMovie.results
        }
    }

    fun setItemClickListener(listener: ItemClickListener) {
        this.listener = listener
    }
}
